### PR TITLE
Block affine component

### PR DESCRIPTION
**Pull request by galv**
Note the commit message regarding a change to nnet-test-utils.cc

There is no actual change in natural-gradient-online-test.cc. I simply accidentally committed a change, and couldn't figure out how to remove it via rebasing, so I had to change it again back to its previous state in the second commit.

### CODE REVIEW

File: src/nnet3/nnet-simple-component.h [hunk 1]
```diff
@@ -454,6 +455,76 @@ class RepeatedAffineComponent: public UpdatableComponent {
...
+  //   N
+  //   O ] but we actually treat it as the matrix:
+  // [ M 0 0
+  //   0 N 0
+  //   0 0 O ]
+  CuMatrix<BaseFloat> linear_params_;
+  CuVector<BaseFloat> bias_params_;
+  int32 num_blocks_;
+ private:
+  const RepeatedAffineComponent &operator = (const RepeatedAffineComponent &other); // Disallow.
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.h]: should be BlockAffineComponent not RepeatedAffineComopnent

File: src/nnet3/nnet-test-utils.cc [hunk 2]
```diff
@@ -1026,12 +1026,21 @@ static void GenerateRandomComponentConfig(std::string *component_type,
     case 24: {
       *component_type = "RepeatedAffineComponent";
       int32 num_repeats = RandInt(1, 50),
-          input_dim = num_repeats * RandInt(1, 15),
-          output_dim = num_repeats * RandInt(1, 15);
```
Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: was there a reason this crashed with 1?

**Comment by danpovey** [comment]
Thanks!  Apart from those very small issues I pointed out, this looks great.
Dan

File: src/nnet3/nnet-test-utils.cc (hunk 2 above)
Comment (galv) [inline src/nnet3/nnet-test-utils.cc]: The numerical routines (i.e., Propagate and Backprop) did not crash when input_dim=1 and output_dim=1. Rather, the crash in ./nnet-component-test was because predicted_obj_change was not approximately equal to measured_objf_change in TestSimpleComponentDataDerivative().

Obviously, the test works for other (larger) input parameters. Simply removing a test case that doesn't work seems a little bad, but I think the test is likely to be a false positive since the tests work for other tested parameters.

Of course I also don't really understand the tests as well as I should. (I don't understand why the output_deriv-output dot product is a valid objective function.)

Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: It prints out the measured and predicted objf changes- how different were
they?  Were they at least usually the same sign?   If they were just a bit
out of the tolerance it's OK to disable that test case, but if they were
completely wrong, it could indicate a problem e.g. with the derivatives for
the bias.
Dan

Comment (galv) [inline src/nnet3/nnet-test-utils.cc]: I'll look into the details of what exactly triggers the failure and get
some numbers soon.

On Thu, Dec 31, 2015 at 3:35 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/nnet3/nnet-test-utils.cc
> <https://github.com/kaldi-asr/kaldi/pull/427#discussion_r48670460>:
>
> > @@ -1026,12 +1026,21 @@ static void GenerateRandomComponentConfig(std::string *component_type,
> >      case 24: {
> >        *component_type = "RepeatedAffineComponent";
> >        int32 num_repeats = RandInt(1, 50),
> > -          input_dim = num_repeats * RandInt(1, 15),
> > -          output_dim = num_repeats * RandInt(1, 15);
>
> It prints out the measured and predicted objf changes- how different were
> they? Were they at least usually the same sign? If they were just a bit out
> of the tolerance it's OK to disable that test case, but if they were
> completely wrong, it could indicate a problem e.g. with the derivatives for
> the bias. Dan
> … <#151f9c09c98656f2_>
> On Thu, Dec 31, 2015 at 9:48 AM, Daniel Galvez <notifications@github.com>
> wrote: In src/nnet3/nnet-test-utils.cc <#427 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/427#discussion_r48664776>>: > @@
> -1026,12 +1026,21 @@ static void GenerateRandomComponentConfig(std::string
> *component_type, > case 24: { > *component_type =
> "RepeatedAffineComponent"; > int32 num_repeats = RandInt(1, 50), > -
> input_dim = num_repeats * RandInt(1, 15), > - output_dim = num_repeats *
> RandInt(1, 15); The numerical routines (i.e., Propagate and Backprop) did
> not crash when input_dim=1 and output_dim=1. Rather, the crash in
> ./nnet-component-test was because predicted_obj_change was not
> approximately equal to measured_objf_change in
> TestSimpleComponentDataDerivative(). Obviously, the test works for other
> (larger) input parameters. Simply removing a test case that doesn't work
> seems a little bad, but I think the test is likely to be a false positive
> since the tests work for other tested parameters. Of course I also don't
> really understand the tests as well as I should. (I don't understand why
> the output_deriv-output dot product is a valid objective function.) — Reply
> to this email directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/427/files#r48664776>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/427/files#r48670460>.
>



-- 
Daniel Galvez
danieltgalvez.com

File: src/nnet3/nnet-simple-component.cc [hunk 3]
```diff
@@ -1291,6 +1292,293 @@ void RepeatedAffineComponent::UnVectorize(const VectorBase<BaseFloat> &params) {
...
+
+    { // linear params update
+
+      std::vector<CuSubMatrix<BaseFloat> *> in_value_batch,
+        out_deriv_batch, linear_params_deriv_batch;
+
+      // this constructor zeros out the entire matrix. Unfortunately, we need
+      // this for AddMatMatBatched even though beta is 0 because it is possible
+      // there may be NaN in the matrix's space, and 0 * NaN is NaN.
+      CuMatrix<BaseFloat> linear_params_deriv(linear_params_.NumRows(),
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.cc]: I don't think the temporary is necessary here.  It was needed in the RepeatedAffineComponent to avoid a race condition, but here you can directly add to linear_params_.  The learning rate is normally large enough that any roundoff error is negligible, and anyway we do it like this elsewhere.

File: src/nnet3/nnet-simple-component.cc [hunk 4]
```diff
@@ -1291,6 +1292,293 @@ void RepeatedAffineComponent::UnVectorize(const VectorBase<BaseFloat> &params) {
...
+     !cfl->GetValue("output-dim", &output_dim) ||
+     !cfl->GetValue("num-blocks", &num_blocks) || cfl->HasUnusedValues()) {
+    KALDI_ERR << "Invalid initializer for layer of type "
+              << Type() << ": \"" << cfl->WholeLine() << "\"";
+  }
+
+  // optional parameters
+  BaseFloat learning_rate = learning_rate_;
+  cfl->GetValue("learning-rate", &learning_rate);
+  BaseFloat param_stddev = 1.0 / std::sqrt(input_dim),
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.cc]: Please make the default param_stddev 1.0 / std::sqrt(input_dim / num_blocks), and also do the same
in the RepeatedAffineComponent.  There's a reason for this (it's natural if you follow the logic in the original paper that introduced this rule, to this case).  I already fixed it in the branch where I'm using it, but not upstream.  It does matter.

File: src/nnet3/nnet-simple-component.cc (hunk 3 above)
Comment (galv) [inline src/nnet3/nnet-simple-component.cc]: I think you're absolutely right here. I can't believe I didn't realize this myself.

File: src/nnet3/nnet-simple-component.cc (hunk 4 above)
Comment (galv) [inline src/nnet3/nnet-simple-component.cc]: I assume you're talking about Glorot intialization. If so, I agree with these changes.

File: src/nnet3/nnet-test-utils.cc (hunk 2 above)
Comment (galv) [inline src/nnet3/nnet-test-utils.cc]: Okay, I have confirmed that the tests fail only RepeatedAffineComponent (NOT BlockAffineComponent. That looks fine.), when input_dim / num_repeats and output_dim / num_repeats are equal to 1. Even then, it is not guaranteed to fail, however. It can also fail for other small values.

I suspect the reason has to do with striding. I'm worried your hack to view the vector as a matrix assumes that does not take into account the effect of extra padding used between rows for the sake of alignment. A padding issue exactly like the kind of thing that would come up only with very small dimensions. I don't know CUDA, but the [documentation](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g32bd7a39135594788a542ae72217775c) mentions that padding is used in allocations.

I'll test this hypothesis soon, hopefully tomorrow.

Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: I think I see what might be the problem.  The following code:
      int32 size_as_vector =
          (linear_params_.NumRows() - 1) * linear_params_.Stride() +
          linear_params_.NumCols(),
          stride_as_matrix = linear_params_.NumRows() *
linear_params_.Stride();
where it says linear_params_.Stride(), needs to refer to
linear_params_deriv_repeated.Stride().
They are usually but not always the same.
Dan

**Comment by galv** [comment]
I've integrated all of your changes and rebased against chain.

I noticed that in the chain branch, you only partially made the striding fix you proposed. You replaced only one instance of linear_params_.Stride() with linear_params_deriv_repeated.Stride(). I made the change in both places, and have thoroughly tested to make sure the gradient checks pass for the case that input-dim= 1* num_repeats_ and output-dim = 1 * num_repeats_, and for small repeated blocks in general.

The previous error must have been due to accessing illegal memory. Sometimes when the downhill test failed, it reported that the new objective function was nan or infinity.

**Comment by danpovey** [comment]
Thanks a lot!   Merging.
